### PR TITLE
Implemented Random and Advanced AI

### PR DESCRIPTION
**Pull request by Marc313**
Other than AI, this branch also includes:
- Dice Probability Calculations
--> Relies on Math functions I added for factorial and combinations

### CODE REVIEW

File: scripts/game/gamemode/roll_phase.gd [hunk 1]
```diff
@@ -5,13 +5,16 @@ extends Phase
...
+	# If its an ai's turn, let the ai know it can 'click' on the dice.
+	if (_gamemode.is_ai_turn()):
+		_gamemode.get_current_ai().roll()
+	else:
+		_gamemode.dice.enable_selection()
+		
 
 ## Turns off dice selection highlight effects.
 func end():
-	_gamemode.dice.disable_selection()
```
Comment (zemiguel20) [inline scripts/game/gamemode/roll_phase.gd:13]: This function guarantees that by the end of the roll phase, the dice are no longer clickable. Should be kept here.

File: scripts/game/gamemode/move_phase.gd [hunk 2]
```diff
@@ -16,13 +22,18 @@ func start():
...
+		_gamemode.get_current_ai().make_move(_legal_moves)
+	else:
+		_link_highlighting()
+		_enable_piece_selection()
 
 
 ## Disables highlighting of the movable pieces calculated in [method start] 
 func end():
-	_unlink_highlighting()
+	if (not _is_ai_turn):
```
Comment (zemiguel20) [inline scripts/game/gamemode/move_phase.gd:35]: Small tip: no need for ( ). Doesnt really need change as it still works, just wanted to point out.

File: Scripts/game/dice_probabilities.gd
Comment (zemiguel20) [inline Scripts/game/dice_probabilities.gd:1]: Theoretically, this functions could be used in other contexts, but in this case they are only here as helper functions for the AI, so I think it should probably be moved to the `ai` folder, to cause less confusion.

**Review by zemiguel20 — CHANGES_REQUESTED** [review]
Looks fine in general. Probably, code will need general cleanup later.
Also try to fix the renaming issue.

File: scripts/game/gamemode/move_phase.gd (hunk 2 above)
Comment (Marc313) [reply scripts/game/gamemode/move_phase.gd:35]: Yeah I am a bit inconsistent with this, i'll try sticking to not using brackets in the future!

**Review by zemiguel20 — APPROVED** [review]

**Pull request merged**